.github/workflows/ci.yml

```diff
@@ -0,0 +1,58 @@
+name: CI
+
+on:
+  push:
+    branches: ["main"]
+  pull_request:
+    branches: ["main"]
+  workflow_dispatch:
+
+env:
+  MIX_ENV: test
+
+jobs:
+  test_macos:
+    name: Elixir ${{ matrix.pair.elixir }} OTP ${{ matrix.pair.otp }} (macOS)
+    runs-on: macos-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        pair:
+          - elixir: "1.13"
+            otp: "24.3.4.10"
+          - elixir: "1.17"
+            otp: "27.0.1"
+    steps:
+      - uses: actions/checkout@v4
+      - name: Generate .tool-versions file
+        run: |
+          echo "elixir ${{ matrix.pair.elixir }}" >> .tool-versions
+          echo "erlang ${{ matrix.pair.otp }}" >> .tool-versions
+          cat .tool-versions
+      - uses: asdf-vm/actions/install@v3
+      - name: Install Hex package manager
+        run: mix local.hex --force
+      - name: Install dependencies
+        run: mix deps.get
+      - run: mix format --check-formatted
+      - run: mix test
+  test_ubuntu:
+    name: Elixir ${{ matrix.pair.elixir }} OTP ${{ matrix.pair.otp }} (Ubuntu)
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        pair:
+          - elixir: "1.13"
+            otp: "24.3.4.10"
+          - elixir: "1.17"
+            otp: "27.0.1"
+    steps:
+      - uses: actions/checkout@v4
+      - uses: CargoSense/setup-elixir-project@v1
+        with:
+          otp-version: ${{ matrix.pair.otp }}
+          elixir-version: ${{ matrix.pair.elixir }}
+          build-flags: "--warnings-as-errors"
+      - run: mix format --check-formatted
+      - run: mix test
```

Comment:
The `lint` portion of the matrix is there to ensure certain command/option combinations (e.g. `mix format`, `mix test --warnings-as-errors`) are run only on the latest Elixir/OTP pair.

```suggestion
      matrix:
        include:
          - pair:
              elixir: "1.13"
              otp: "24.3.4.10"
          - pair:
              elixir: "1.17"
              otp: "27.0.1"
            lint: lint
    steps:
      - uses: actions/checkout@v4
      - uses: erlef/setup-beam@main
        with:
          otp-version: ${{ matrix.pair.otp }}
          elixir-version: ${{ matrix.pair.elixir }}
          version-type: strict
      - uses: actions/cache@v4
        with:
          path: deps
          key: mix-deps-${{ hashFiles('**/mix.lock') }}
      - run: mix deps.get --check-locked
      - run: mix format --check-formatted
        if: ${{ matrix.lint }}
      - run: mix deps.unlock --check-unused
        if: ${{ matrix.lint }}
      - run: mix deps.compile
      - run: mix compile --warnings-as-errors
        if: ${{ matrix.lint }}
      - run: mix test
        if: ${{ ! matrix.lint }}
      - run: mix test --warnings-as-errors
        if: ${{ matrix.lint }}
```

Note I'm also recommending we skip our custom GitHub Action in favor of `erlef/setup-beam`. The phoenix_playground repo has a simplified config that should work fine for this repo:
- https://github.com/phoenix-playground/phoenix_playground/blob/main/.github/workflows/ci.yml#L1